tests/libvmi/vmi.go

```diff
@@ -135,3 +135,35 @@ func baseVmi(name string) *v1.VirtualMachineInstance {
 	}
 	return vmi
 }
+
+// WithResourceCpu specifies the vmi cpu resource.
+func WithResourceCPU(value string) Option {
+	return func(vmi *v1.VirtualMachineInstance) {
+		if vmi.Spec.Domain.Resources.Requests == nil {
+			vmi.Spec.Domain.Resources.Requests = k8sv1.ResourceList{}
+		}
+		vmi.Spec.Domain.Resources.Requests[k8sv1.ResourceCPU] = resource.MustParse(value)
+	}
+}
+
+// WithSerialBios if set the BIOS output will be transmitted over serial.
+func WithSerialBios() Option {
+	return func(vmi *v1.VirtualMachineInstance) {
+		if vmi.Spec.Domain.Firmware == nil {
+			vmi.Spec.Domain.Firmware = &v1.Firmware{}
+		}
+		if vmi.Spec.Domain.Firmware.Bootloader == nil {
+			vmi.Spec.Domain.Firmware.Bootloader = &v1.Bootloader{}
+		}
+		if vmi.Spec.Domain.Firmware.Bootloader.BIOS == nil {
+			vmi.Spec.Domain.Firmware.Bootloader.BIOS = &v1.BIOS{}
+		}
+		vmi.Spec.Domain.Firmware.Bootloader.BIOS.UseSerial = pointer.Bool(true)
+	}
+}
+
+func WithClientPassthrough() Option {
```

Comment:
What is the value of making it local if we will probably use it again?

Reply:
@iholder-redhat  
I would argue the same for `WithSEV`. It should be in `sev.go` where it would be pretty self-contained.

`WithNodeSelectorFor` seems to have a usage in multiple locations in the test suite (although it is not yet directly used)and seems to be generic enough to reside in libvmi directly.


`WithClientPassthrough`, `WithSerialBios` seems like one-off options. We would just make a new pile of stuff(similar to utils file which is being witch-hunted) if we would be adding these options here.


@Barakmor1 
I think we can look for an analogy with functions. You don't make functions public just for the sake we might want to use them in the future. You make them private and you can make them public if the needs arise.

Reply:
> If libvmi is supposed to fit many use cases and many users I don't see why not have a "pile of stuff" as long as it's generic and not coupled to a specific use-case. If we implement such an option that probably would fit external users - why make them reinvent the same wheel?

The other side of this is that typically, it is hard to define a good API for many use cases when you have only a single data point at hand. Hence it is prudent to keep things local until enough evidence show that the suggested API is good. In this case, it is not clear at all if something very specific as whether the VM bios speak over serial line should bother each reader of the main part of `libvmi` or should it sit in another file? If we do not know now what is the reasonable structure, we can postpone the decision to the day where the second user of the function shows up.

Reply:
> The other side of this is that typically, it is hard to define a good API for many use cases when you have only a single data point at hand. Hence it is prudent to keep things local until enough evidence show that the suggested API is good. 

The way I see it, since we're in the position of designing a public library that is expected to be used by other users, but we have no such users yet, we have to take some sophisticated guesses. After all, there is no evidence for any API yet :)

> In this case, it is not clear at all if something very specific as whether the VM bios speak over serial line should bother each reader of the main part of `libvmi` or should it sit in another file?

I'm fine with keeping the main part of libvmi small, we could simply add separate files for specific configurations (e.g. firmware). That way we would keep the main file clean but still will provide many options for external users.

> If we do not know now what is the reasonable structure, we can postpone the decision to the day where the second user of the function shows up.

:+1: 

I think we should decide on a general approach moving forward. If we need to provide evidence for adding functions to libvmi, we should define what a "good evidence" is while we have no external users to guide us.

Reply:
I attempted to provide an approximate definition: add a public only after you have two independent use cases for it.

It is very hard to predefine what a good API looked like. In healthy projects there is a tension between maintainers who want to keep a small bugless API and multiple users with conflicting requests to extend the API. A good API arises gradually from this tension.

Reply:
> add a public API only after you have two independent use cases for it.

Again, in my eyes we are just another user. The ways we use these APIs should not set our design decisions imho.

> It is very hard to predefine what a good API looked like. In healthy projects there is a tension between maintainers who want to keep a small bugless API and multiple users with conflicting requests to extend the API. A good API arises gradually from this tension.

These options basically set a field into a struct. I don't think that there are much bugs to expose with such trivial implementation. And there are no users, so no conflicting requests, or any requests, exist. As said before, I think we would have to take guesses here. There is no other way imho. 

There's a very thin line between an internal library and a public library that is designed solely by how we are using it.

-----

Also, IMHO, this is a general discussion that needs to be done outside of the context of this PR. There are many other options (e.g. `WithSEV()`) that are already public. I think it would be much better to define what a "good API" is, what are the standards of adding a public option, etc, in a broader context.

If we change this in this PR only we end up in a mixed position and no clear path forward.

But all I can do is raise my protest :)

Reply:
I think that we need a well defined convention that include this topic and also for the matter of static funcs definitions.
otherwise these rules will be violated / these Discussions will repeat them self  .

Reply:
> As said before, I think we would have to take guesses here. There is no other way imho.

There is a way: keep the function local and do not commit on an API  until a second user shows up.

> But all I can do is raise my protest :)

No need for protest. If you feel sure enough that a certain function already belong in the library API, just say say so in the commit message and convince the library maintainer. It is a very thin line, and I don't think a clear-cut definition exists.

> also for the matter of static funcs definitions

Here, I believe, a clear-cut definition *does* exist. Lambdas are very cool and very powerful, but they incur cognitive burden and often too powerful. They force me to read their implementation when all I care is about their usage. Hence I strongly request to avoid them unless you really need their power (e.g need to change the function implementation on the fly).

Reply:
Still I think we can define conventions for usage of Lambdas (without enforcement).
I think that you have a point and the usage of Lambda in here was wrong but without
conventions this will not last as every reviewer have a different opinion
and there are many violations of the conclusions we talked about.

Reply:
> Still I think we can define conventions for usage of Lambdas (without enforcement). I think that you have a point and the usage of Lambda in here was wrong but without conventions this will not last as every reviewer have a different opinion and there are many violations of the conclusions we talked about.

I am aware that we have many needlessly-confusing lambdas (I was just tipped about pile of them in [operator_test](https://github.com/kubevirt/kubevirt/blob/main/tests/operator_test.go#L113)). I think that a KubeVirt-wide convention against them would make a lot of sense.

Reply:
> > As said before, I think we would have to take guesses here. There is no other way imho.
> 
> There is a way: keep the function local and do not commit on an API until a second user shows up.

So we shouldn't expose any options since there is no second user for any of them?
 
> > But all I can do is raise my protest :)
> 
> No need for protest. If you feel sure enough that a certain function already belong in the library API, just say say so in the commit message and convince the library maintainer. It is a very thin line, and I don't think a clear-cut definition exists.

Right, but which factors make the "library maintainer" convinced?
I really think we should discuss and define these terms.

Right now I do think that the functions presented in this PR are general enough to be exposed. So I express it and try to convince. But it's hard to do so when we don't have any clear policy / guidelines or path forward.

Reply:
> Here, I believe, a clear-cut definition does exist. Lambdas are very cool and very powerful, but they incur cognitive burden and often too powerful. They force me to read their implementation when all I care is about their usage. Hence I strongly request to avoid them unless you really need their power (e.g need to change the function implementation on the fly).

BTW, I don't agree with this either :)

I don't understand why a lambda functions "force me to read their implementation". Who's forcing you? :smile: Why are you forced to read the implementation of lambdas and not regular functions? Just skip the implementation and go straight to the usage, similar to any other function.

Defining a helper-lambda-function within another function lets you bound the context and scope of the function. This usually makes sense IMO when the function is meaningless outside of that context, which is pretty common for helper functions.

That being said, the fact that we have a thousand lambdas per test in situations that make no sense is clearly a bad habit.

Reply:
> The difference is that lambda are defined in-place. Good tests are short. REALLY short. You glance at them and you know what they are doing. When I see a test that is calling a function named withSerialBIOS() that's all I need to know. If the function is defined in-scope, it hides more important things. The most important purpose of functions is to define an abstraction, which makes us think faster about the important things. This is challenged is the abstracted complexity is pushed to the face of the reader.

Let me give an example and let's see your opinion on it :)

```go
Context("VM with specific firmware", func() {

    setInvalidFirmware := func() {...}
    moreVeryContextSpecificHelperFunc := func() {...}

    It("ensure invalid firmware is rejected, func() {
        vm := ...
        setInvalidFirmware(vm)
        err := virtClient.Create(vm)
        Expect(err).To(HaveOccurred())
    })

})
```

In my opinion this usage is completely valid and encouraged. The advantages are:
- isFirmwareValid() has no meaning outside of this test. No one needs to set an invalid firmware outside of this specific context. Moreover, it would be very confusing to expose this function, even as a "private" function.
- The test is still very readable and straight forward.

Do you agree?

Reply:
Sorry @Barakmor1 for hijacking your PR. Since we agree that your new functions are meaningful in the file context, we can take our discussion elsewhere.

> Do you agree?

I'm not sure that I do. But I'm willing to rescind my statement about the case against lamdas being "clear cut". There are exceptions.

I see no awkwardness in a file-local function named `withInvalidFirmware` and I find it hard to image what `moreVeryContextSpecificHelperFunc` may be. When the function implementation is hidden as `{...}` it does not distract me as much when I read the including context. I think that we can make better progress if you show me a real-life kubevirt-code named lambda that does not change, does not use context variable, yet improves readability. I suspect that there are such functions, but I bet we would find them in too-long too-cluttered files.

Reply:
> Sorry @Barakmor1 for hijacking your PR. Since we agree that your new functions are meaningful in the file context, we can take our discussion elsewhere.

sounds good :+1: 
This is indeed beyond the scope of this PR :)

> 
> > Do you agree?
> 
> I'm not sure that I do. But I'm willing to rescind my statement about the case against lamdas being "clear cut". There are exceptions.
> 
> I see no awkwardness in a file-local function named `withInvalidFirmware` and I find it hard to image what `moreVeryContextSpecificHelperFunc` may be. When the function implementation is hidden as `{...}` it does not distract me as much when I read the including context. I think that we can make better progress if you show me a real-life kubevirt-code named lambda that does not change, does not use context variable, yet improves readability. I suspect that there are such functions, but I bet we would find them in too-long too-cluttered files.

I'm up for the challenge :)
And as said before, I think that you're right for most cases :) It's a great rule of thumb, but like all rules of thumb, they aren't correct for every situation imho

p.s. a modern IDE can "collapse" function implementations. So in my eyes these functions are always hidden behind `{...}`.